### PR TITLE
 Add a setting to ignore dirty tabs in workbench.editor.limit.value.

### LINKED ISSUE #144309: Add a setting to ignore dirty tabs in `workbench.editor.limit.value`
I have long been looking for a way to keep my tabs tidy and was extremely happy to find the `workbench.editor.limit` feature.
Per @[bpasero](https://github.com/bpasero) https://github.com/microsoft/vscode/issues/9872#issuecomment-567463617, dirty tabs count towards the editor tabs limit. This is unfortunate, because I often have several dirty tabs and once their number approaches the open tabs limit, I no longer get to open several clean tabs in parallel. I would like to have an option for `workbench.editor.limit.value` to ignore dirty tabs.

### PR DESCRIPTION
**Pull request by luoriyuhui21**
This PR fixes #144309

### CODE REVIEW

**Comment by luoriyuhui21** [comment]
This PR fixes #144545

File: src/vs/workbench/browser/parts/editor/editorsObserver.ts [hunk 1]
```diff
@@ -312,6 +312,7 @@ export class EditorsObserver extends Disposable {
 		}
 
 		const limit = this.editorGroupsService.partOptions.limit.value;
+		const ignoreDirtyTab: boolean = this.editorGroupsService.partOptions.limit.ignoreDirtyTab ? this.editorGroupsService.partOptions.limit.ignoreDirtyTab : false;
```
Comment (bra1nDump) [inline src/vs/workbench/browser/parts/editor/editorsObserver.ts:315]:
```suggestion
		const ignoreDirtyTab: boolean = this.editorGroupsService.partOptions.limit.ignoreDirtyTab ?? false;
```

File: src/vs/workbench/browser/parts/editor/editorsObserver.ts [hunk 2]
```diff
@@ -359,9 +360,17 @@ export class EditorsObserver extends Disposable {
 
 			return true;
 		});
+		let editorsToCloseCount: number;
+		if (ignoreDirtyTab === true) {
+			editorsToCloseCount = leastRecentlyClosableEditors.length - limit + 1;
```
Comment (bra1nDump) [inline src/vs/workbench/browser/parts/editor/editorsObserver.ts:365]: Why the +1?
```suggestion
			editorsToCloseCount = leastRecentlyClosableEditors.length - limit;
```

**Review by bra1nDump — COMMENTED** [review]
Another name I would consider is ignoreDirityTabs. There are multiple tabs we can ignore.

File: src/vs/workbench/browser/parts/editor/editorsObserver.ts (hunk 1 above)
Comment (luoriyuhui21) [reply src/vs/workbench/browser/parts/editor/editorsObserver.ts:315]: OK， Maybe “ignoreDirityTabs” is  proper

File: src/vs/workbench/browser/parts/editor/editorsObserver.ts (hunk 2 above)
Comment (luoriyuhui21) [reply src/vs/workbench/browser/parts/editor/editorsObserver.ts:365]: I set  Workbench.editor.limit to 5 and enable ignoreDirtyTab,   and I have 3 dirty files ,then I can only open 4 files when without "+1";  and add "+1", I can open  5 files.   The reson should be that leastRecentlyClosableEditors don't include the sticky tab

**Comment by luoriyuhui21** [comment]
@bra1nDump  Is there any problem about this PR？

**Comment by bra1nDump** [comment]
@luoriyuhui21 I don't have write rights to this repo. You need to wait for someone from vscode team to review it.

I just came across your pull request in 'good first issues' read it and added my 2 cents :D

File: src/vs/workbench/common/editor.ts [hunk 3]
```diff
@@ -968,6 +968,7 @@ interface IEditorPartConfiguration {
 	splitOnDragAndDrop?: boolean;
 	limit?: {
 		enabled?: boolean;
+		ignoreDirtyTabs?: boolean;
```
Comment (bpasero) [inline src/vs/workbench/common/editor.ts:971]: Suggest `excludeDirty`

File: src/vs/workbench/browser/workbench.contribution.ts [hunk 4]
```diff
@@ -252,6 +252,11 @@ const registry = Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Con
 				'exclusiveMinimum': 0,
 				'markdownDescription': localize('limitEditorsMaximum', "Controls the maximum number of opened editors. Use the `#workbench.editor.limit.perEditorGroup#` setting to control this limit per editor group or across all groups.")
 			},
+			'workbench.editor.limit.ignoreDirtyTabs': {
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.contribution.ts:255]: Suggest `workbench.editor.limit.excludeDirty`

File: src/vs/workbench/browser/workbench.contribution.ts [hunk 5]
```diff
@@ -252,6 +252,11 @@ const registry = Registry.as<IConfigurationRegistry>(ConfigurationExtensions.Con
...
+				'type': 'boolean',
+				'default': false,
+				'description': localize('limitEditorsIgoreDirtyTabs', "Controls if the maximum number of opened editors should ignore dirty tabs. When enabled, ignore the dirty tabs")
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.contribution.ts:258]: Suggest: `Controls if the maximum number of opened editors should exclude dirty editors for the configured limit.`

File: src/vs/workbench/browser/parts/editor/editorsObserver.ts [hunk 6]
```diff
@@ -312,6 +312,7 @@ export class EditorsObserver extends Disposable {
 		}
 
 		const limit = this.editorGroupsService.partOptions.limit.value;
+		const ignoreDirtyTabs: boolean = this.editorGroupsService.partOptions.limit.ignoreDirtyTabs ?? false;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorsObserver.ts:315]: Suggest `excludeDirty`

File: src/vs/workbench/browser/parts/editor/editorsObserver.ts [hunk 7]
```diff
@@ -359,9 +360,17 @@ export class EditorsObserver extends Disposable {
 
 			return true;
 		});
+		let editorsToCloseCount: number;
+		if (ignoreDirtyTabs === true) {
+			editorsToCloseCount = leastRecentlyClosableEditors.length - limit + 1;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorsObserver.ts:365]: I am not sure this logic is entirely correct. The list of `leastRecentlyClosableEditors` is not only driven by dirty editors but also sticky editors (sticky editors are the "Pinned" editors in the UI). There is also another check for `never the editor that should be excluded`. I think the check for dealing with dirty editors or not needs to go somewhere closer where we check for dirty editors.

**Review by bpasero — CHANGES_REQUESTED** [review]

File: src/vs/workbench/browser/workbench.contribution.ts (hunk 4 above)
Comment (luoriyuhui21) [reply src/vs/workbench/browser/workbench.contribution.ts:255]: OK

File: src/vs/workbench/browser/parts/editor/editorsObserver.ts (hunk 7 above)
Comment (luoriyuhui21) [reply src/vs/workbench/browser/parts/editor/editorsObserver.ts]: @bpasero Thank you for your suggestions, I changed the logic about editorsToCloseCount , I feel it's more clear now， please review it.

**Comment by bpasero** [comment]
@luoriyuhui21 I took a slightly different approach and have a `mostRecentEditorsCountingForLimit` to make it really obvious that the setting reduces the set of editors that are considered for the limit, can you please test this change works fine for you?

Thanks!

**Comment by luoriyuhui21** [comment]
https://github.com/microsoft/vscode/pull/144545

> @luoriyuhui21 I took a slightly different approach and have a [`mostRecentEditorsCountingForLimit`](https://github.com/microsoft/vscode/pull/144545/commits/33fc92cb2279c31d980c2ba1956366db5fd58375#diff-0b29aa58b928762137f69e4cb19a50ec9679d3d288a005702891215f8578f7daR345) to make it really obvious that the setting reduces the set of editors that are considered for the limit, can you please test this change works fine for you?
> 
> Thanks!

OK, I test your changes in my project. It works fine in vscode. 
And your changes  make code more easy to read, which is better. it is ok for me, thank you.

**Pull request merged**